### PR TITLE
Add operational style api for egressed artifacts

### PR DESCRIPTION
**Pull request by wiktork**
Note this follows a similar approach as the Azure Sdk (https://docs.microsoft.com/en-us/azure/azure-maps/creator-long-running-operation)

Some design considerations/possible changes
- Status and error return the same status code but slightly different structure. We could use our own 2xx code here to make it easier for callers to determine the state.
- There is no bounds on the operations we track. They stay around forever.
- 202 Accepted is a good return value but can be easily missed by the client. E.g. curl with egress without a -v just shows nothing.
- We could consider a redirect for the above as well.

Remaining items:
- [ ] Add more logging and comments
- [x] Consider a name pass over all the newly introduced types
- [x] Unit tests
- [x] Account for infinite duration tasks
- [ ] Update documentation

### CODE REVIEW

File: src/Tools/dotnet-monitor/DiagnosticsMonitorCommandHandler.cs
Comment (jander-msft) [inline src/Tools/dotnet-monitor/DiagnosticsMonitorCommandHandler.cs:231]: I think we should create a "add everything that the rest server implements" extension method and make all of these an implementation detail.

File: src/Microsoft.Diagnostics.Monitoring.RestServer/Operation/EgressOperationStore.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.RestServer/Operation/EgressOperationStore.cs:16]: I this interface really necessary? Isn't this an internal implementation detail of the server assembly?

File: src/Microsoft.Diagnostics.Monitoring.RestServer/Operation/EgressOperationStore.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.RestServer/Operation/EgressOperationStore.cs:72]: This and the below methods should not be Task-returning methods if they aren't async, if the service is just an implementation within the server assembly.

File: src/Microsoft.Diagnostics.Monitoring.RestServer/Operation/EgressOperationQueue.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.RestServer/Operation/EgressOperationQueue.cs:14]: Is this interface needed? It looks like this is an implementation detail of the server assembly.

File: src/Microsoft.Diagnostics.Monitoring.RestServer/Operation/EgressOperationQueue.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.RestServer/Operation/EgressOperationQueue.cs:36]: nit: EnqueueAsync

File: src/Microsoft.Diagnostics.Monitoring.RestServer/Controllers/OperationController.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.RestServer/Controllers/OperationController.cs:42]: Let's not assume JSON. Try using `ControllerBase.StatusCode(int, object)`

File: src/Microsoft.Diagnostics.Monitoring.RestServer/Controllers/OperationController.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.RestServer/Controllers/OperationController.cs:57]: Can you add a comment that CancelOperation will throw an exception if the operation ID doesn't exist? Reading it as-is, one might think it would return (and thus return 200) regardless of the existence of the operation ID.

**Review by jander-msft — CHANGES_REQUESTED** [review]

File: src/Microsoft.Diagnostics.Monitoring.RestServer/Operation/EgressOperationStore.cs
Comment (wiktork) [reply src/Microsoft.Diagnostics.Monitoring.RestServer/Operation/EgressOperationStore.cs:16]: It's not that I'm creating a public api here, this is just standard dependency injection into the controller. I can use the class directly, but it will need to remain public.

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperationService.cs
Comment (wiktork) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperationService.cs:41]: This approach will not work; it forces serialization of all requests, and indefinite requests will block all other requests from the queue.

File: src/Tests/test.runsettings
Comment (jander-msft) [inline src/Tests/test.runsettings:2]: Let's just delete this file. I don't think setting this environment variable helps at all.

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperationService.cs
Comment (jander-msft) [reply src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperationService.cs:41]: Maybe you can create another async method that effectively does a Task.WhenAny on the in-flight executions + one task from a completion source that is signaled when when a new operation is dequeued. If that last task is signaled, just loop back around and start a new Task.WhenAny. If any other task, complete the operation.

Another option is to add continuations to each of the tasks that completes the operation when it finishes.

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/EgressTests.cs
Comment (wiktork) [reply src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/EgressTests.cs:99]: I didn't see an easy way to hook configuration for the tool. I can see if it's easy to add.

File: documentation/openapi.json
Comment (kkeirstead) [inline documentation/openapi.json:749]: For consistency with the new format, should this be something like "/operation?{operationId}" instead? Looking at the rest of the PR, I'm assuming this hasn't been merged with the main branch in a while since other endpoints are still using the old format, but just something I wanted to make sure I pointed out :)

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/OperationController.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/OperationController.cs:26]: nit: Should this be "operations" since it represents the entire collection of operations?

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/HttpApi/ApiClient.cs
Comment (kkeirstead) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/HttpApi/ApiClient.cs:325]: Just another case of using the old format -> "/trace?pid={processId}&..."

Comment (wiktork) [reply src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/HttpApi/ApiClient.cs:325]: Yep, I didn't fix these up after merging.

File: documentation/openapi.json
Comment (wiktork) [reply documentation/openapi.json:749]: I actually intend to leave these as they are. I don't think these are the same types of requests as our diagnostic based requests, and follow REST semantics more closely. I.e. an operation has a unique id, where as a process is not (and there are several ways to specify a process).

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressResult.cs
Comment (wiktork) [reply src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressResult.cs:9]: I would like to keep it in case we end up with some more data here at some point.

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/DiagController.cs
Comment (wiktork) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/DiagController.cs:718]: Note this does not work as expected for netcore 3.1 (it works fine for 5.0). It appears that StatusCodeResult gets converted to ObjectResult, and then there's a content type mismatch. This produces a 406.

Comment (wiktork) [reply src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/DiagController.cs:718]: Thinking about this some more, I think we are going up against the declared api controller behavior, which is to produce a problemdetails. My tentative thinking is to get rid of the middleware altogether and let the exception throw and then get converted to an appropriate problemdetails.

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Models/EgressOperationStatus.cs
Comment (kkeirstead) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/Models/EgressOperationStatus.cs:33]: I don't fully understand the relationship, so it might not be appropriate, but would it make sense for OperationStatus to inherit from OperationSummary (since the summary should always be a subset of the full status)? The duplication of the getters and setters between the classes caught my eye, so I thought I'd mention it. Also, relatedly, am I correct in assuming Id is the same as OperationId? In which case I'd think they should probably be named the same, and if not maybe a comment to point out the difference?

Comment (wiktork) [reply src/Microsoft.Diagnostics.Monitoring.WebApi/Models/EgressOperationStatus.cs:33]: I thought about doing this but I really wanted to enforce that a new object had to be made here. I'll correct the names and add some comments.

File: src/Microsoft.Diagnostics.Monitoring.WebApi/RequestThrottling/RequestLimitTracker.cs
Comment (wiktork) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/RequestThrottling/RequestLimitTracker.cs:43]: nit: Localize

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/EgressTests.cs
Comment (wiktork) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/EgressTests.cs:99]: Add another test for listing all operations

File: src/Microsoft.Diagnostics.Monitoring.WebApi/ActionContextExtensions.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/ActionContextExtensions.cs:115]: nit: Maybe create a static instance and return that so long as ExecutionResult is immutable.

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperationStore.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperationStore.cs:68]:
```suggestion
            await _taskQueue.EnqueueAsync(request).ConfigureAwait(false);
```

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperationService.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperationService.cs:53]:
```suggestion
                var result = await egressRequest.EgressOperation.ExecuteAsync(_serviceProvider, token).ConfigureAwait(false);
```

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperationService.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperationService.cs:36]: No need to await.

```suggestion
                _ = Task.Run(() => ExecuteEgressOptions(egressResult, stoppingToken));
```

if you do keep await, add `.ConfigureAwait(false)`

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperation.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/Operation/EgressOperation.cs:41]: Please add `.ConfigureAwait(false);`

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/DiagController.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/DiagController.cs:707]:
```suggestion
            Guid operationId = await _operationsStore.AddOperation(egressStreamResult, limitKey).ConfigureAwait(false);
```

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/DiagController.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/DiagController.cs:228]: Add `.ConfigureAwait(false)`

File: src/Microsoft.Diagnostics.Monitoring.WebApi/ActionContextExtensions.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/ActionContextExtensions.cs:118]:
```suggestion
            ExecutionResult<object> result = await ExecutionHelper.InvokeAsync(innerAction, logger, context.HttpContext.RequestAborted).ConfigureAwait(false);
```

File: src/Microsoft.Diagnostics.Monitoring.WebApi/ActionContextExtensions.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/ActionContextExtensions.cs:121]:
```suggestion
                await context.ProblemAsync(new BadRequestObjectResult(result.ProblemDetails)).ConfigureAwait(false);
```

File: src/Microsoft.Diagnostics.Monitoring.WebApi/ActionContextExtensions.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/ActionContextExtensions.cs:30]: nit: Is this an extra newline before the method body?

File: src/Microsoft.Diagnostics.Monitoring.WebApi/ActionContextExtensions.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/ActionContextExtensions.cs:17]: nit: It may be better to make this immutable. Also, maybe add two static methods for `static ExecutionResult<T> Success<T>(T result)` and `static ExecutionResult<T> Failed<T>(ProblemDetails details)`

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Models/EgressOperationStatus.cs
Comment (jander-msft) [inline src/Microsoft.Diagnostics.Monitoring.WebApi/Models/EgressOperationStatus.cs:30]: Please remove newline so that openapi document doesn't have `\r\n` in it.

**Review by jander-msft — CHANGES_REQUESTED** [review]

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/HttpApi/ApiClient.cs
Comment (jander-msft) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/HttpApi/ApiClient.cs:320]: This method should be enforcing successful before returning, something like:

```cs
switch (response.StatusCode)
{
    case HttpStatusCode.OK:
        ValidateContentType(response, ContentTypes.ApplicationJson);
        return new OperationResponse(response.StatusCode, response.Headers.Location, await response.Content.ReadAsStringAsync().ConfigureAwait(false));
    case HttpStatusCode.BadRequest:
    case HttpStatusCode.TooManyRequests:
        ValidateContentType(response, ContentTypes.ApplicationProblemJson);
        throw await CreateValidationProblemDetailsExceptionAsync(response).ConfigureAwait(false);        
    case HttpStatusCode.Unauthorized:
        ThrowIfNotSuccess(response);
        break;
}

throw await CreateUnexpectedStatusCodeExceptionAsync(response).ConfigureAwait(false);
```

and the test callers should be catching the exceptions when those are expected (e.g. when testing the throttling).

All of the new ApiClient methods should be do this.

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/HttpApi/ApiClient.cs
Comment (jander-msft) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/HttpApi/ApiClient.cs:353]: Why is this necessary? The models used by the tests should be exactly the same as used by the product, so there should be no disagreement of casing.

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/Options/OptionsExtensions.cs
Comment (jander-msft) [inline src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/Options/OptionsExtensions.cs:118]: I don't think this change is necessary anymore due to the egress configuration refactor.

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/OperationController.cs
Comment (jander-msft) [reply src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/OperationController.cs:27]: You resolved this comment, but I don't see any action or follow up on it. Are you planning on making this change?

**Review by jander-msft — APPROVED** [review]
I have a few more things I'd like for you to change mostly in the tests, but I'm not going to block the PR anymore.

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/OperationController.cs
Comment (wiktork) [reply src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/OperationController.cs:27]: I think I misunderstood the scope of your comment: you want to rename the entire controller not just the field.

File: src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/HttpApi/ApiClient.cs
Comment (wiktork) [reply src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/HttpApi/ApiClient.cs:353]: I will annotate the model to reflect the default serialization style instead of doing this.

File: src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/OperationController.cs
Comment (jander-msft) [reply src/Microsoft.Diagnostics.Monitoring.WebApi/Controllers/OperationController.cs:27]: Correct, I would expect the routes to be `/operations` and `/operations/{id}` to insinuate that I'm listing all of the operations and that I'm getting a single operation by id from the collection of operations, respectively.

**Pull request merged**
